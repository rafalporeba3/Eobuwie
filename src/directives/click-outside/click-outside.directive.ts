import { DirectiveOptions, VNode } from 'vue';
import { DirectiveBinding }        from 'vue/types/options';
import { Vue }                     from 'vue/types/vue';

let handleOutsideClick!: any;

export const clickOutsideDirective: DirectiveOptions = {
  bind(el: HTMLElement, binding: DirectiveBinding, vnode: VNode) {
    handleOutsideClick = (e: MouseEvent) => {
      e.stopPropagation();
      const { handler, excludeElements } = binding.value;
      let clickedOnExcludedEl = false;

      excludeElements.forEach((refName: string) => {
        if (!clickedOnExcludedEl) {
          const excludedEl: Element = (vnode?.context?.$refs[ refName ] as Vue)?.$el;
          clickedOnExcludedEl = excludedEl.contains(e.target as any);
        }
      });

      if (!el.contains(e.target as any) && !clickedOnExcludedEl && vnode && vnode.context) {
        vnode.context[ handler as keyof Vue ]();
      }
    };
    document.addEventListener('click', handleOutsideClick);
    document.addEventListener('touchstart', handleOutsideClick);
  },

  unbind() {
    document.removeEventListener('click', handleOutsideClick);
    document.removeEventListener('touchstart', handleOutsideClick);
  },
};
