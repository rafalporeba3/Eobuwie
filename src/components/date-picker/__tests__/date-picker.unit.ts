import { createLocalVue, Wrapper }   from '@vue/test-utils';
import Vuex                          from 'vuex';
import CompositionApi                from '@vue/composition-api';
import { createTestInstanceShallow } from '@/__tests__/factory';
import { clickOutsideDirective }     from '@/directives/click-outside/click-outside.directive';
import cDatePicker                   from '../date-picker.component.vue';
import cDatePickerWrapper            from '../components/wrapper/date-picker-wrapper.component.vue';
import cDatePickerEditor             from '../components/editor/date-picker-editor.component.vue';
import cDatePickerPanel              from '../components/panel/date-picker-panel.component.vue';

const localVue = createLocalVue();
localVue.use(CompositionApi);
localVue.use(Vuex);
localVue.directive('click-outside', clickOutsideDirective);

describe('Date picker component', () => {
  let wrapper!: Wrapper<any>;

  beforeEach(() => {
    const { wrapperInstance } = createTestInstanceShallow(cDatePicker, {
      localVue,
      propsData: {
        options: {},
        selectedDates: [],
        disabledDates: [],
      },
    });

    wrapper = wrapperInstance;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render cDatePicker component with correct class', () => {
    expect(wrapper.classes())
      .toContain('c-date-picker');
  });

  it('should contain core child components', () => {
    expect(wrapper.findComponent(cDatePickerWrapper)
      .exists())
      .toBeTruthy();
    expect(wrapper.findComponent(cDatePickerEditor)
      .exists())
      .toBeTruthy();
    expect(wrapper.findComponent(cDatePickerPanel)
      .exists())
      .toBeTruthy();
  });

  it('should update panelVisibility on cDatePickerEditor click', async () => {
    const datePickerEditor: Wrapper<any> = wrapper.findComponent(cDatePickerEditor);
    jest.spyOn(wrapper.vm, 'changePanelVisibility');

    await datePickerEditor.trigger('click');

    expect(wrapper.vm.changePanelVisibility)
      .toHaveBeenCalled();

    expect(wrapper.vm.isPanelVisible)
      .toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(wrapper.html())
      .toMatchSnapshot();
  });
});
