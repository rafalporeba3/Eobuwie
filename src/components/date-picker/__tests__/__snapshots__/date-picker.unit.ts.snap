// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date picker component should match snapshot 1`] = `
<c-date-picker-wrapper-stub class="c-date-picker">
  <c-date-picker-editor-stub startplaceholder="Check In" endplaceholder="Check Out"></c-date-picker-editor-stub>
  <c-date-picker-panel-stub style="display: none;"></c-date-picker-panel-stub>
</c-date-picker-wrapper-stub>
`;
